edb/pgsql/compiler/dml.py

```diff
@@ -1206,13 +1283,22 @@ def insert_needs_conflict_cte(
     if on_conflict.always_check or ir_stmt.conflict_checks:
         return True
 
+    # We can't use ON CONFLICT if there are access policies
+    # on the type, since UNLESS CONFLICT only should avoid
+    # conflicts with objects that are visible.
     type_id = ir_stmt.subject.typeref.real_material_type.id
     if (
         (type_id, True) in ctx.env.type_rewrites
         or (type_id, False) in ctx.env.type_rewrites
     ):
         return True
 
+    # We can't use ON CONFLICT if there are rewrites on the type
+    # because rewrites might reference multi pointers, which means
+    # we need to execute link operations before the final INSERT.
+    if ir_stmt.rewrites and ir_stmt.rewrites.by_type:
+        return True
```

Comment:
What happens if there are ON CONFLICT and rewrites? Shouldn't the result of `merge_overlays_globally` be checked and an error thrown when True?

Reply:
If we can't use SQL's `ON CONFLICT`, then we do conflict detection manually, and that's fine.

... unless the UNLESS CONFLICT is actually on a rewritten pointer, in which case I'm pretty sure we'll do the wrong thing, because the conflict checks on INSERT are compiled in a way that duplicates the original expressions in the set. (UNLESS CONFLICT doesn't work on volatile properties for this reason. I think we have the infrastructure that would be required to mostly fix this, though.)

For now we should probably disallow using UNLESS CONFLICT on a rewritten property though